### PR TITLE
PR: Load files in the background and set focus to the last one with focus in the previous session

### LINKED ISSUE #3585: Editor grabs focus when opening files at startup
## Description of your problem

**What steps will reproduce the problem?**
1. In spyder open a couple of files in editor
2. Close spyder and restart
3. 

**What is the expected output? What do you see instead?**
 After the main window appears, it is grabbing focus for each file that is reopened from the previous session. In addition each of these files is put to the front of the editor window. 

Reopening files in a thread is a good idea to reduce startup time. However, at the moment this is not effective, as the user will not be able to do anything (neither in spyder nor in any other application) until all files are loaded.

**Please provide any additional information below**
## Versions and main components
- Spyder Version: 4.0.0dev 858193f
- Python Version: 2.7.12
- Operating system: osx 10.10

### PR DESCRIPTION
**Pull request by mariacamilaremolinagutierrez**
Fixes #3585

### CODE REVIEW

**Comment by ccordoba12** [comment]
@mariacamilaremolinagutierrez, I changed this PR to be against master, as we talked about it last night :-)

File: spyder/plugins/editor.py [hunk 1]
```diff
@@ -1839,38 +1841,72 @@ def _convert(fname):
...
+                    self.__add_recent_file(filename)
+                if goto is not None: # 'word' is assumed to be None as well
+                    current_editor.go_to_line(goto[index], word=word)
+                    position = current_editor.get_position('cursor')
+                    self.cursor_moved(filename0, position0, filename, position)
+                current_editor.clearFocus()
+                current_editor.setFocus()
+                current_editor.window().raise_()
+                if processevents:
+                    QApplication.processEvents()
```
Comment (ccordoba12) [inline spyder/plugins/editor.py]: There's a lot of repeated code in this `if/else` block. Please simplify it.

File: spyder/plugins/editor.py [hunk 2]
```diff
@@ -2617,6 +2627,25 @@ def setup_open_files(self):
             self.__load_temp_file()
         self.set_create_new_file_if_empty(True)
 
+    def reorder_filenames(self, filenames):
```
Comment (ccordoba12) [inline spyder/plugins/editor.py]: Please add a docstring to this function to know what it does

File: spyder/widgets/editor.py [hunk 3]
```diff
@@ -2090,11 +2092,14 @@ def set_layout_settings(self, settings):
...
-                except IndexError:
+                if dont_goto is not None:
+                    # skip go to line for first file because is already there
                     pass
-            editorstack.set_current_filename(cfname)
+                else:
+                    try:
+                        editor.go_to_line(clines[index])
+                    except IndexError:
+                        pass
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: Is this the part that needs fixing in a second PR?

Comment (mariacamilaremolinagutierrez) [inline spyder/widgets/editor.py]: Yes this is the part!

Comment (ccordoba12) [inline spyder/widgets/editor.py]: Ok, then please leave a `TODO` comment here. saying that this is going to be fixed in a future PR :-)

Comment (ccordoba12) [inline spyder/widgets/editor.py]: Also mention in this comment the reason why this needs to be fixed.

Comment (mariacamilaremolinagutierrez) [inline spyder/widgets/editor.py]: @ccordoba12 This is ready

**Comment by ccordoba12** [comment]
@mariacamilaremolinagutierrez, please stop merging with 3.x or master until is absolutely necessary (i.e. there are merge conflicts) or we advise you to do so.

Otherwise there's no need for you to do a merge :-)

File: spyder/widgets/editor.py [hunk 4]
```diff
@@ -2089,12 +2091,17 @@ def set_layout_settings(self, settings):
             editorstack = splitter.widget(0)
             for index, finfo in enumerate(editorstack.data):
                 editor = finfo.editor
-                # FIXME: Temporal fix
-                try:
-                    editor.go_to_line(clines[index])
-                except IndexError:
+                # TODO: go_to_line is not working properly (the line it jumps
+                # is not the corresponding to that file). It will be fix in a
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: `... It will be fixed in a ...`

File: spyder/widgets/editor.py [hunk 5]
```diff
@@ -2089,12 +2091,17 @@ def set_layout_settings(self, settings):
             editorstack = splitter.widget(0)
             for index, finfo in enumerate(editorstack.data):
                 editor = finfo.editor
-                # FIXME: Temporal fix
-                try:
-                    editor.go_to_line(clines[index])
-                except IndexError:
+                # TODO: go_to_line is not working properly (the line it jumps
+                # is not the corresponding to that file). It will be fix in a
+                # future PR (fixing issue #3857)
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: `future PR (which will fix issue #3857)`

File: spyder/widgets/editor.py [hunk 6]
```diff
@@ -2089,12 +2091,17 @@ def set_layout_settings(self, settings):
             editorstack = splitter.widget(0)
             for index, finfo in enumerate(editorstack.data):
                 editor = finfo.editor
-                # FIXME: Temporal fix
-                try:
-                    editor.go_to_line(clines[index])
-                except IndexError:
+                # TODO: go_to_line is not working properly (the line it jumps
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: `... the line it jumps to ...`

File: spyder/widgets/editor.py (hunk 4 above)
Comment (ccordoba12) [inline spyder/widgets/editor.py]: Sorry, this whole line should read as

`does not correspond to the proper file). This will be fixed in a ...`

**Comment by ccordoba12** [comment]
This looks to me, I just made some grammar comments about the TODO you added.

Just please remember to make the next PR against master and not 3.x :-)

File: spyder/plugins/editor.py [hunk 7]
```diff
@@ -1839,27 +1841,32 @@ def _convert(fname):
             goto = [goto]
         elif goto is not None and len(goto) != len(filenames):
             goto = None
-            
+
         for index, filename in enumerate(filenames):
             # -- Do not open an already opened file
-            current_editor = self.set_current_filename(filename, editorwindow)
+            if index == 0:  # this is the last file focused in previous session
+                foc = True
```
Comment (ccordoba12) [inline spyder/plugins/editor.py]: Why not name this variable as `focus`?

File: spyder/plugins/editor.py [hunk 8]
```diff
@@ -2617,6 +2627,33 @@ def setup_open_files(self):
             self.__load_temp_file()
         self.set_create_new_file_if_empty(True)
 
+    def reorder_filenames(self, filenames):
+        """Take the last session filenames and puts the last open on first.
```
Comment (ccordoba12) [inline spyder/plugins/editor.py]: `... and put the last one open first.`

**Comment by mariacamilaremolinagutierrez** [comment]
@ccordoba12 Thank you Carlos, this is ready :)

File: spyder/widgets/editor.py [hunk 9]
```diff
@@ -2092,8 +2092,8 @@ def set_layout_settings(self, settings, dont_goto=None):
             for index, finfo in enumerate(editorstack.data):
                 editor = finfo.editor
                 # TODO: go_to_line is not working properly (the line it jumps
-                # is not the corresponding to that file). It will be fix in a
-                # future PR (fixing issue #3857)
+                # to is not the corresponding to that file). This will be fix
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: `... This will be fixed`

File: spyder/widgets/editor.py [hunk 10]
```diff
@@ -2089,12 +2091,17 @@ def set_layout_settings(self, settings):
             editorstack = splitter.widget(0)
             for index, finfo in enumerate(editorstack.data):
                 editor = finfo.editor
-                # FIXME: Temporal fix
-                try:
-                    editor.go_to_line(clines[index])
-                except IndexError:
+                # TODO: go_to_line is not working properly (the line it jumps
+                # to is not the corresponding to that file). This will be fix
```
Comment (ccordoba12) [inline spyder/widgets/editor.py]: `... This will be fixed`

**Comment by mariacamilaremolinagutierrez** [comment]
@ccordoba12 This one is ready :)

**Comment by ccordoba12** [comment]
Ok, let's merge it so you can fix the goto line problems :-)

**Pull request merged**